### DISCUSSION TITLE
A CPP code example for the RP6502

### DISCUSSION THREAD

**Discussion by vruumllc** [picocomputer/community#84] [Q&A]
Here is a non-trivial CPP example for the RP6502 compiled with llvm-mos.
It seems to work great!
It's a little demo of a text-mode framework I'm working on. 
It is still very much a work-in-progress, and my CPP is a little rusty, so be kind 8^)

[txt_test_cpp.zip](https://github.com/picocomputer/community/files/14254661/txt_test_cpp.zip)

**Comment by rumbledethumps** [comment]
Yay! I no longer have to declare local variables at top of scope!

**Reply by vruumllc** [reply]
I’m super excited about llvm-mos. Can you tell? 8^)

**Comment by vruumllc** [comment]
IGNORE:
I tried to write a text editor for the RP6502 using llvm-mos C++, but ran out of memory before I could finish.
I then tried repackaging nearly identical code into C, and found that it used less than 1/2 the memory (23k vs 53k)!!!

The editor is still not done, but I thought the llvm-mos folks might find this interesting. 
(Also anyone interested in creating a modern, mouse-aware, text application for the RP6502).

This is not to bash the llvm-mos C compiler -- it ROCKS! I love it.

Enjoy! tonyvr

[te_cpp.zip](https://github.com/picocomputer/community/files/14397119/te_cpp.zip)

EDIT: DON'T USE THIS! IT IS COMPILED IN DEBUG MODE. SEE BELOW
[te_c.zip](https://github.com/picocomputer/community/files/14397120/te_c.zip)

**Reply by rumbledethumps** [reply]
Be sure to try the MinSizeRelease build. LLVM-MOS likes to inline and unroll things.

**Reply by vruumllc** [reply]
PLEASE IGNORE THIS TOO (I think I need more caffeine):
I'm pretty sure I am using that, but I will try rebuilding llvm-mos myself. I managed to do it for my ARM64 Pinebook-Pro, which I use on the road, so I should be able to get it done for my Intel laptop.

In any case, I used the same llvm-mos package for both projects, so the results are still relevant.

**Reply by vruumllc** [reply]
ARGH!! It looks like the C++ project somehow got switched into DEBUG mode! I changed it to MinSizeRelease mode, and it is now 29k.
Please ignore my comments above!!! CPP is still viable.  I am totally thrilled (if embarrassed).  Her is a properly compiled C++ version:

OK, both versions are compiled in MinSizeRelease mode now:
[te_cpp.zip](https://github.com/picocomputer/community/files/14397197/te_cpp.zip)
[te_c.zip](https://github.com/picocomputer/community/files/14397224/te_c.zip)

**Reply by vruumllc** [reply]
Well, at least there is now a C version, in case anyone wants to compare llvm-mos with cc65 (or they just don't like C++). 8^)

**Reply by vruumllc** [reply]
It just occurred to me that this text mode framework could be used for hybrid bitmap/text apps if I switched the plane for the menus to plane1, and use plane0 for bitmap mode. You could then make a real paint program, for instance, and have all the menu support to save and load files already to go in text.
